compiler/rustc_mir/src/transform/large_enums.rs

```diff
@@ -0,0 +1,214 @@
+use crate::transform::MirPass;
+use crate::util::patch::MirPatch;
+use rustc_middle::mir::*;
+use rustc_middle::ty::{self, Const, List, ParamEnv, TyCtxt};
+use rustc_target::abi::{Size, Variants};
+
+/// A pass that seeks to optimize unnecessary moves of large enum types, if there is a large
+/// enough discrepanc between them
+pub struct EnumSizeOpt<const DISCREPANCY: u64>;
+
+impl<'tcx, const D: u64> MirPass<'tcx> for EnumSizeOpt<D> {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        self.optim(tcx, body);
+    }
+}
+
+impl<const D: u64> EnumSizeOpt<D> {
+    fn optim(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        for l in body.vars_and_temps_iter() {
+            let l_decl = &mut body.local_decls[l];
+            let (num_variants, sizes, match_ty): (u64, Vec<_>, _) = match l_decl.ty.kind() {
+                ty::Adt(adt_def, _substs) if adt_def.is_enum() => {
+                    let p_e = tcx.param_env(adt_def.did);
+                    // TODO handle error better below
+                    let variants = &tcx.layout_of(p_e.and(l_decl.ty)).unwrap().variants;
+                    match variants {
+                        Variants::Single { .. } => continue,
+                        Variants::Multiple { variants, .. } if variants.len() <= 1 => continue,
+                        Variants::Multiple { variants, .. } => {
+                            let min = variants.iter().map(|v| v.size).min().unwrap();
+                            let max = variants.iter().map(|v| v.size).max().unwrap();
+                            if max.bytes() - min.bytes() < D {
+                                continue;
+                            }
+                            (
+                                variants.len() as u64,
+                                variants.iter().map(|v| v.size).collect(),
```

Comment:
Since you're `collect`ing the sizes at the end anyway:
```suggestion
                            let sizes = variants.iter().map(|v| v.size).collect();
                            let min = sizes.min().unwrap();
                            let max = sizes.max().unwrap();
                            if max.bytes() - min.bytes() < D {
                                continue;
                            }
                            (
                                variants.len() as u64,
                                sizes,
```

Reply:
It might not always allocate if the threshold is not met on line 32, so I think collecting early might be a bit slower than waiting. I think this is also a bit of a premature optimization, but will keep in mind for later.

Reply:
Yeah, agreed. Had the same thought about the continue branch in the mean time but got distracted from editing my comment.
Although the suggestion came more from a readability angle than from an optimization angle.